### PR TITLE
[Fix] Register optimizer constructor with mmseg

### PR DESCRIPTION
**Pull request by linfangjian01**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix to register optimizer constructor directly into mmcv.

## Modification

1.  Add mmseg/core/builder.py to register OPTIMIZER_BUILDERS of mmcv.
2.  Change the build_optimizer import from mmcv to mmseg in mmseg/apis/train.py.
3.  Change the OPTIMIZER_BUILDERS import from mmcv to mmseg in mmseg/core/utils/layer_decay_optimizer_constructor.py and mmseg/core/layer_decay_optimizer_constructor.py.

### CODE REVIEW

**Comment by linfangjian01** [comment]
This pr is not perfect and is still under modification.

**Comment by MeowZheng** [comment]
> ###

Please revise the tag [Fix] -> [WIP], as WIP (working in process) means this pr is not ready to review

File: mmseg/core/utils/layer_decay_optimizer_constructor.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 import json
 
-from mmcv.runner import (OPTIMIZER_BUILDERS, DefaultOptimizerConstructor,
-                         get_dist_info)
+from mmcv.runner import DefaultOptimizerConstructor, get_dist_info
 
 from ...utils import get_root_logger
```
Comment (MeowZheng) [inline mmseg/core/utils/layer_decay_optimizer_constructor.py:6]:
```suggestion
from mmseg.utils import get_root_logger
```

File: tests/test_core/test_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
+    paramwise_cfg = dict()
+    optim_constructor_cfg = dict(
+        type='DefaultOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    optimizer = optim_constructor(model)
+    # Test whether optimizer constructor can be built from parent.
+    assert isinstance(optimizer, torch.optim.SGD)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:54]:
```suggestion
    optimizer_cfg = dict(
        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
    paramwise_cfg = dict()
    optim_constructor_cfg = dict(
        type='DefaultOptimizerConstructor',
        optimizer_cfg=optimizer_cfg,
        paramwise_cfg=paramwise_cfg)
    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
    # Test whether optimizer constructor can be built from parent.
    assert isinstance(optim_constructor, DefaultOptimizerConstructor)
```
```suggestion
    model = ExampleModel()
    optimizer_cfg = dict(
        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
    paramwise_cfg = dict()
    optim_constructor_cfg = dict(
        type='DefaultOptimizerConstructor',
        optimizer_cfg=optimizer_cfg,
        paramwise_cfg=paramwise_cfg)
    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
    optimizer = optim_constructor(model)
    # Test whether optimizer constructor can be built from parent.
    assert isinstance(optimizer, torch.optim.SGD)
```

Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:54]:
```suggestion
    optimizer_cfg = dict(
        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
    paramwise_cfg = dict()
    optim_constructor_cfg = dict(
        type='DefaultOptimizerConstructor',
        optimizer_cfg=optimizer_cfg,
        paramwise_cfg=paramwise_cfg)
    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
    # Test whether optimizer constructor can be built from parent.
    assert isinstance(optim_constructor, DefaultOptimizerConstructor)
```
```suggestion
    model = ExampleModel()
    optimizer_cfg = dict(
        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
    paramwise_cfg = dict()
    optim_constructor_cfg = dict(
        type='DefaultOptimizerConstructor',
        optimizer_cfg=optimizer_cfg,
        paramwise_cfg=paramwise_cfg)
    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
    optimizer = optim_constructor(model)
    # Test whether optimizer constructor can be built from parent.
    assert isinstance(optimizer, torch.optim.SGD)
```

File: tests/test_core/test_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    paramwise_cfg = dict(conv1_lr_mult=5)
+    optim_constructor_cfg = dict(
+        type='MyOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    optimizer = optim_constructor(model)
+    # Tests whether a new optimizer constructor can be registered.
+    assert isinstance(optimizer, torch.optim.SGD)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:86]:
```suggestion
    optimizer = optim_constructor(model)
    # Tests whether a new optimizer constructor can be registered.
    assert isinstance(optim_constructor, MyOptimizerConstructor)
```

File: tests/test_core/test_optimizer.py (hunk 2 above)
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:54]:
```suggestion
    optimizer_cfg = dict(
        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
    paramwise_cfg = dict()
    optim_constructor_cfg = dict(
        type='DefaultOptimizerConstructor',
        optimizer_cfg=optimizer_cfg,
        paramwise_cfg=paramwise_cfg)
    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
    # Test whether optimizer constructor can be built from parent.
    assert isinstance(optim_constructor, DefaultOptimizerConstructor)
```

File: tests/test_core/test_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+class ExampleModel(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.param1 = nn.Parameter(torch.ones(1))
+        self.conv1 = nn.Conv2d(3, 4, kernel_size=1, bias=False)
+        self.conv2 = nn.Conv2d(4, 2, kernel_size=1)
+        self.bn = nn.BatchNorm2d(2)
+        self.sub = SubModel()
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:31]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+class SubModel(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.conv1 = nn.Conv2d(2, 2, kernel_size=1, groups=2)
+        self.gn = nn.GroupNorm(2, 2)
+        self.param1 = nn.Parameter(torch.ones(1))
+
+    def forward(self, x):
+        return x
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:20]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    optimizer_cfg = dict(
+        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
+    paramwise_cfg = dict()
+    optim_constructor_cfg = dict(
+        type='DefaultOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test whether optimizer constructor can be built from parent.
+    assert isinstance(optim_constructor, DefaultOptimizerConstructor)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:39]:
```suggestion
    assert type(optim_constructor, DefaultOptimizerConstructor)
```

File: tests/test_core/test_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            conv1_lr_mult = self.paramwise_cfg.get('conv1_lr_mult', 1.)
+
+            params = []
+            for name, param in model.named_parameters():
+                param_group = {'params': [param]}
+                if name.startswith('conv1') and param.requires_grad:
+                    param_group['lr'] = self.base_lr * conv1_lr_mult
+                params.append(param_group)
+            optimizer_cfg['params'] = params
+
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:60]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            return build_from_cfg(optimizer_cfg, OPTIMIZERS)
+
+    paramwise_cfg = dict(conv1_lr_mult=5)
+    optim_constructor_cfg = dict(
+        type='MyOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test whether optimizer constructor can be built from parent.
+    assert isinstance(optim_constructor, DefaultOptimizerConstructor)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:70]:
```suggestion
    # Test optimizer constructor can be built from child registry.
    assert not type(optim_constructor, DefaultOptimizerConstructor)
```

File: tests/test_core/test_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    # Tests whether a new optimizer constructor can be registered.
+    assert isinstance(optim_constructor, MyOptimizerConstructor)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:72]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    optim_constructor_cfg = dict(
+        type='DefaultOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test whether optimizer constructor can be built from parent.
+    assert type(optim_constructor) is DefaultOptimizerConstructor
+
+    from mmcv.runner import OPTIMIZERS
+    from mmcv.utils import build_from_cfg
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:42]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    @OPTIMIZER_BUILDERS.register_module()
+    class MyOptimizerConstructor(DefaultOptimizerConstructor):
+
+        def __call__(self):
+
+            return build_from_cfg(optimizer_cfg, OPTIMIZERS)
+
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:50]:
```suggestion
    pass
```

File: tests/test_core/test_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    paramwise_cfg = dict(conv1_lr_mult=5)
+    optim_constructor_cfg = dict(
+        type='MyOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test optimizer constructor can be built from child registry.
+    assert type(optim_constructor) is not DefaultOptimizerConstructor
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:58]: assert type(optim_constructor) is MyOptimizerConstructor

File: tests/test_core/test_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:59]:
```suggestion
# test invalid constructor cannot be built
with pytest.raise(KeyError):
    build_optimizer_constructor(dict(type='A'))
```

File: tests/test_core/test_optimizer.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test whether optimizer constructor can be built from parent.
+    assert type(optim_constructor) is DefaultOptimizerConstructor
+
+    @OPTIMIZER_BUILDERS.register_module()
+    class MyOptimizerConstructor(DefaultOptimizerConstructor):
+
+        pass
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:44]:
```suggestion
    @OPTIMIZER_BUILDERS.register_module()
    class MyOptimizerConstructor(DefaultOptimizerConstructor):
        pass
```

File: tests/test_core/test_optimizer.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    optim_constructor_cfg = dict(
+        type='MyOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test optimizer constructor can be built from child registry.
+    assert type(optim_constructor) is MyOptimizerConstructor
+    # test invalid constructor cannot be built
+    with pytest.raises(KeyError):
+        build_optimizer_constructor(dict(type='A'))
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:56]:
```suggestion

    # Test unregistered constructor cannot be built
    with pytest.raises(KeyError):
        build_optimizer_constructor(dict(type='A'))
```

**Review by MeowZheng — APPROVED** [review]

File: tests/test_core/test_optimizer.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+base_lr = 0.01
+base_wd = 0.0001
+momentum = 0.9
+
+
+def test_build_optimizer_constructor():
+    optimizer_cfg = dict(
+        type='SGD', lr=base_lr, weight_decay=base_wd, momentum=momentum)
+    paramwise_cfg = dict()
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:32]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    optim_constructor_cfg = dict(
+        type='DefaultOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:36]:
```suggestion
        optimizer_cfg=optimizer_cfg)
```

File: tests/test_core/test_optimizer.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    optim_constructor = build_optimizer_constructor(optim_constructor_cfg)
+    # Test whether optimizer constructor can be built from parent.
+    assert type(optim_constructor) is DefaultOptimizerConstructor
+
+    @OPTIMIZER_BUILDERS.register_module()
+    class MyOptimizerConstructor(DefaultOptimizerConstructor):
+        pass
+
+    paramwise_cfg = dict(conv1_lr_mult=5)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:45]:
```suggestion
```

File: tests/test_core/test_optimizer.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    optim_constructor_cfg = dict(
+        type='MyOptimizerConstructor',
+        optimizer_cfg=optimizer_cfg,
+        paramwise_cfg=paramwise_cfg)
```
Comment (MeowZheng) [inline tests/test_core/test_optimizer.py:49]:
```suggestion
        optimizer_cfg=optimizer_cfg)
```

**Pull request merged**